api/types/externalaudit/externalaudit.go

```diff
@@ -0,0 +1,189 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package externalaudit
+
+import (
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/header"
+	"github.com/gravitational/teleport/api/types/header/convert/legacy"
+	"github.com/gravitational/teleport/api/utils"
+)
+
+// ExternalAudit
+type ExternalAudit struct {
+	// ResourceHeader is the common resource header for all resources.
+	header.ResourceHeader
+
+	// Spec is the specification for the external audit.
+	Spec ExternalAuditSpec `json:"spec" yaml:"spec"`
+}
+
+// ExternalAuditSpec is the specification for an external audit.
+type ExternalAuditSpec struct {
+	// IntegrationName is name of existing OIDC intagration used to
+	// generate AWS credentials.
+	IntegrationName string `json:"integration_name" yaml:"integration_name"`
+	// SessionsRecordingsURI is s3 path used to store sessions recordings.
+	SessionsRecordingsURI string `json:"sessions_recordings_uri" yaml:"sessions_recordings_uri"`
+	// AthenaWorkspace is workspace used by Athena audit logs during queries.
+	AthenaWorkspace string `json:"athena_workspace" yaml:"athena_workspace"`
+	// GlueDatabase is database used by Athena audit logs during queries.
+	GlueDatabase string `json:"glue_database" yaml:"glue_database"`
+	// GlueTable is table used by Athena audit logs during queries.
+	GlueTable string `json:"glue_table" yaml:"glue_table"`
+	// AuditEventsLongTermURI is s3 path used to store batched parquet files with
+	// audit events, partitioned by event date.
+	AuditEventsLongTermURI string `json:"audit_events_long_term_uri" yaml:"audit_events_long_term_uri"`
+	// AthenaResultsURI is s3 path used to store temporary results generated by
+	// Athena engine.
+	AthenaResultsURI string `json:"athena_results_uri" yaml:"athena_results_uri"`
+}
+
+// NewExternalAudit will create a new external audit.
+func NewExternalAudit(metadata header.Metadata, spec ExternalAuditSpec) (*ExternalAudit, error) {
+	externalaudit := &ExternalAudit{
+		ResourceHeader: header.ResourceHeaderFromMetadata(metadata),
+		Spec:           spec,
+	}
+
+	if err := externalaudit.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return externalaudit, nil
+}
+
+// CheckAndSetDefaults validates fields and populates empty fields with default values.
+func (a *ExternalAudit) CheckAndSetDefaults() error {
```

Comment:
We should enforce that no expiration is set, or we might end up with no configuration after an auth restart. Some singleton resources do it by just not setting the expiration time on the backend write, but it seems a bit untidy to do for user-managed configuration like `external_audit`.

Reply:
it's reworked and explicitly set to zero on CheckAndSetDefaults